### ISSUE TITLE
Problem on Windows 10

### ISSUE THREAD

**Issue by dirkk0** [sinclairzx81/hammer#4]
Hi,
  first of all - thanks for Hammer! I am using it on MacOSX and Linux with no problem.

On Win10 it fails with this error message:

    node:events:504
          throw er; // Unhandled 'error' event
          ^
    
    Error: spawn mkdir ENOENT

Any idea why that is and how to run it properly?

Thanks, Dirk

**Comment by sinclairzx81** [comment]
@dirkk0 Hi, what Hammer command are you running? 

If you are using `serve` or `run`,  Hammer will create a directory using `mkdir` for your output bundle (the `--dist` directory). The `mkdir` error may suggest a permission issue, or you're running a Node version < 14x.

If it is a permission issue, you may want to check the user account you're running as has appropriate permissions to create directories. You could test by moving your project to another location (for example, your desktop) and running the same command, or checking the permissions on the `--dist` directory.  Note, Hammer should work fine on Windows 10 systems (It was actually developed in a Windows 10 environment). 

Hope that helps
S

**Comment by dirkk0** [comment]
Hi, thanks for the swift reply!
I installed a brand new nodejs (16.14.2) and I ran 'hammer serve' within VS Code which itself is 'run as Administrator'. So I would assume it's not a permission issue.
In fact I just did an 'mkdir test' on the terminal and it worked just fine.

**Comment by dirkk0** [comment]
I now moved the directory to the desktop and the error is still there. I again created a directory with no problems with mkdir.

**Comment by sinclairzx81** [comment]
Hmmm, ok. It's actually quite difficult to narrow down on what the issue might be. 

What was the exact command you were running?

**Comment by dirkk0** [comment]
It was 'hammer serve' - is there a better command to narrow down the problem?

**Comment by sinclairzx81** [comment]
If you're using `serve`, then the easiest way to help me narrow down the problem will be to create a very simple example `repro` project on your desktop.

### Repro

```
/Desktop/repro
/Desktop/repro/index.html
```
Then run the following command from inside the `repro` directory.
```
.../Desktop/repro> hammer serve index.html --dist target
```

Just confirm hammer creates the `target` directory inside `repro`

### Next

If the above works ok, then the problem may be to do with referenced assets found in the HTML (for example `<img>`, `<link>`, `<video>` etc). Hammer will copy these into the `--dist` directory relative to the `index.html` file (which may require directory creation). If you are referencing assets in directories up from the current `index.html` file, this may cause problems (as hammer may try to write these above the `--dist` directory (as relative to the current `index.html` file)). So be mindful of this.

### Finally

If you can help narrow down if its a `dist` creation problem, or a HTML asset copy problem, that would help me provide a solution for you. Unfortunately without knowing much about your project directory layout, it's difficult to provide much assistance, but you should be able to replicate things in the above repro. 

Hope this helps!
S

**Comment by dirkk0** [comment]
I did exactly what you described and the error persists.

Out of curiousity, I additionally tried bot Power Shell and 'normal' command prompt (i.e. windowskey-r) with the command you provided and the error is always the same.

**Comment by dirkk0** [comment]
I now additionally ran the command prompt as an Adminstrator and issued the coomand - same problem.
![image](https://user-images.githubusercontent.com/625394/162566189-730cde45-47f7-4e1c-9cee-0b22ca766fb2.png)
I seem to miss something obvious here, but I just don't see it.

**Comment by sinclairzx81** [comment]
@dirkk0 Yes, that's strange as there is nothing in the stacktrace to indicate the error is actually originating from Hammer (everything looks internal to Node), and the only ChildProcess being spawned would be to ESBuild process, so curious if the error is originating from there (I wouldn't have much control over that).

### Logging

I guess to further narrow down the error, you could drop a `try/catch` around the following code (you will need to track down the install location of hammer). Location: `node_modules/@sinclair/hammer/build/build.js - line 76`

```typescript
    copyAsset(asset) {
        if (!fs.existsSync(asset.sourcePath))
            return;
        const directory = path.dirname(asset.targetPath);
        if (!fs.existsSync(directory))
            fs.mkdirSync(directory, { recursive: true }); // TRY/CATCH/CONSOLE.LOG here
        if (asset.type === 'html') {
            fs.writeFileSync(asset.targetPath, asset.content, 'utf-8');
        }
        else {
            fs.copyFileSync(asset.sourcePath, asset.targetPath);
        }
    }
```
But I'm not convinced hammer is throwing this error.

### Versions

The following are the relevant versions for my current Hammer development environment on Windows 10. If you can spot a disparity, can possibly look at updating Hammer to a fixed version of ESBuild (the semver version is currently Esbuild `^0.14.5`)

```bash
Windows 10 Home: Version 10.0.19043 Build 19043
Node: 16.13.1
@sinclair/hammer: 0.16.3
esbuild: 0.14.31
```
This could be a esbuild version issue, ill do a few tests this side.

**Comment by sinclairzx81** [comment]
@dirkk0 Hmmm, I still can't seem to reproduce the issue this side. The following are the steps I used to install a fresh global instance of Hammer. The following commands were run inside an Administrator Powershell Prompt and go through the motions of creating the repro + index.html, installing hammer and running the `serve` command. Everything should be working ok.

```powershell
PS C:\.....\Desktop> mkdir repro


    Directory: C:\.....\Desktop


Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d-----          4/9/2022   7:11 PM                repro


PS C:\.....\Desktop> cd repro
PS C:\.....\Desktop\repro> $null > index.html
PS C:\.....\Desktop\repro> ls


    Directory: C:\...\Desktop\repro


Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a----          4/9/2022   7:11 PM              2 index.html


PS C:\...\Desktop\repro> npm install @sinclair/hammer -g

changed 3 packages, and audited 4 packages in 1s

found 0 vulnerabilities
PS C:\.....\Desktop\repro> hammer
Version: 0.16.3

Commands:

   $ hammer run     <script path>     <...options>
   $ hammer build   <file or folder>  <...options>
   $ hammer watch   <file or folder>  <...options>
   $ hammer serve   <file or folder>  <...options>
   $ hammer monitor <file or folder>  <shell command>
   $ hammer task    <name>            <...arguments>
   $ hammer version
   $ hammer help

Options:

   --target      targets     The es build targets.
   --platform    platform    The target platform.
   --dist        directory   The target directory.
   --port        port        The port to listen on.
   --external    packages    Omits external packages.
   --esm                     Use esm module target.
   --minify                  Minifies the output.
   --sourcemap               Generate sourcemaps.
   --sabs                    (serve) Enable shared array buffer.
   --cors                    (serve) Enable cors.

Done: 8ms
PS C:\...\Desktop\repro> hammer serve index.html
[serve]: C:\...\Desktop\repro\dist on 5000
PS C:\...\Desktop\repro> ls


    Directory: C:\...\Desktop\repro


Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d-----          4/9/2022   7:12 PM                dist
-a----          4/9/2022   7:11 PM              2 index.html
```
Checking the installation directory, Hammer is installed with the following packages and their respective versions.
```
Windows 10 Home:    Version 10.0.19043 Build 19043
Node:               16.13.1

@sinclair/hammer:   0.16.3
esbuild:            0.14.34
esbuild-windows-64: 0.14.34
```

Outside of this, I'm not sure if I can be much more help here mate. I guess you could try and reinstall Node. I am curious to know what the issue might be, if it comes down to some incompatible Node / Esbuild conflict, I would be keen to document this somewhere. Unfortunately, without being able to reproduce this issue, there's not much more assistance I can offer.

If you manage to figure out what the issue is, please let me know (so I can either document or mitigate for other users). Ill leave this issue open for a few days as reference.
Cheers
S

**Comment by dirkk0** [comment]
Thank you so much for your help!

I tried several things like rebooting in safe mode (somebody thought virus protection could be the culprit) which wasn't helping and I created a directory with fs without problems.

I'l check other options and try to go the PowerShell route like you did. As soon as I'll find something out, I'll let you know.

Thanks again!

**Comment by dirkk0** [comment]
ok, I found out what the problem was. And yes, this is embarrassing. :)
I installed hammer with `npm install hammer -g` which - *drumroll* - is a completely different library.

Installing it the proper way (as it is in the documentation! if you read it, at least!) with 
`npm install @sinclair/hammer -g`
'fixes' the installation.

So thanks @sinclairzx81  for writing this explicitely into your latest answer which then finally caught my attention, after all.

Closing this and I hope this helps somebody else.
